packages/devtools_app/lib/src/debugger/file_search.dart

```diff
@@ -124,72 +123,187 @@ class FileSearchFieldState extends State<FileSearchField>
+      if (file[matchIndex] == query[queryIndex]) {
+        final start = matchIndex;
+        final end = matchIndex + 1;
+        autoCompleteResultSegments.add(Range(start, end));
+        queryIndex++;
       }
     }
+    return autoCompleteResultSegments;
+  }
 
-    return AutoCompleteMatch(
-      match,
-      matchedSegments: autoCompleteResultSegments,
-    );
+  String _getFileName(String fullPath) {
+    return _fileNamesCache[fullPath] ??= fullPath.split('/').last;
   }
+}
 
-  List<ScriptRef> _findMatches(
-    String query,
-    List<ScriptRef> scriptRefs,
-  ) {
+class FileSearchResults {
+  FileSearchResults.emptyQuery(this.allScripts) : query = FileQuery.empty();
+
+  FileSearchResults.withQuery({
+    @required this.query,
+    @required this.allScripts,
+  }) {
+    // ignore: prefer_asserts_in_initializer_lists
+    assert(!query.isEmpty);
+
+    for (final scriptRef in allScripts) {
+      if (query.isExactFullPathMatch(scriptRef)) {
+        _exactFullPathMatches.add(scriptRef);
+      } else if (query.isFuzzyMatch(scriptRef)) {
+        _fuzzyMatches.add(scriptRef);
+      }
+    }
+  }
+
+  FileSearchResults._({
+    @required this.query,
+    @required this.allScripts,
+    @required exactFullPathMatches,
+    @required fuzzyMatches,
+  }) {
+    _exactFullPathMatches.addAll(exactFullPathMatches);
+    _fuzzyMatches.addAll(fuzzyMatches);
+  }
+
+  final List<ScriptRef> allScripts;
+  final FileQuery query;
+  final List<ScriptRef> _exactFullPathMatches = [];
+  final List<ScriptRef> _fuzzyMatches = [];
+
+  FileSearchResults get topMatches => _buildTopMatches();
+
+  List<ScriptRef> get scriptRefs =>
+      query.isEmpty ? allScripts : [..._exactFullPathMatches, ..._fuzzyMatches];
+
+  List<AutoCompleteMatch> get autoCompleteMatches => query.isEmpty
+      ? allScripts.map((script) => AutoCompleteMatch(script.uri)).toList()
+      : [
+          ..._exactFullPathMatches
+              .map(query.createExactFullPathAutoCompleteMatch)
+              .toList(),
+          ..._fuzzyMatches
+              .map(query.createFuzzyMatchAutoCompleteMatch)
+              .toList(),
+        ];
+
+  FileSearchResults _buildTopMatches() {
     if (query.isEmpty) {
-      return scriptRefs;
+      return FileSearchResults.emptyQuery(
+        allScripts.sublist(
+          0,
+          numOfMatchesToShow,
+        ),
+      );
     }
 
-    final exactMatches = [];
-    final fuzzyMatches = [];
-
-    for (final scriptRef in scriptRefs) {
-      final fullPath = scriptRef.uri;
-      final fileName =
-          _fileNamesCache[scriptRef.uri] ??= scriptRef.uri.split('/').last;
-      if (fullPath.caseInsensitiveContains(query)) {
-        exactMatches.add(scriptRef);
-      } else if (fileName.caseInsensitiveFuzzyMatch(query)) {
-        fuzzyMatches.add(scriptRef);
-      }
+    if (scriptRefs.length <= numOfMatchesToShow) {
+      // Make a copy of this:
+      return FileSearchResults._(
+        allScripts: allScripts,
```

Comment:
you can make a `copyWith` method that defaults to all of the values passed in here and allows you to optionally pass in something different. You could use this below where you pass 
`    exactFullPathMatches: topMatches[0],
      fuzzyMatches: topMatches[1],`

Reply:
Good idea - switched to using `copyWith`